wrappers/javascript/packages/anoncreds-shared/src/api/W3CCredential.ts

```diff
@@ -0,0 +1,231 @@
+        linkSecret: options.linkSecret,
+        revocationRegistryDefinition
+      })
+
+      // We can discard previous handle and store the new one
+      this.handle.clear()
+      this.handle = credential
+    } finally {
+      objectHandles.forEach((handle) => {
+        handle.clear()
+      })
+    }
+    return this
+  }
+
+  public get schemaId() {
+    return anoncreds.w3cCredentialGetAttribute({ objectHandle: this.handle, name: 'schema_id' })
+  }
+
+  public get credentialDefinitionId() {
+    return anoncreds.w3cCredentialGetAttribute({ objectHandle: this.handle, name: 'cred_def_id' })
+  }
+
+  public get revocationRegistryId() {
+    return anoncreds.w3cCredentialGetAttribute({ objectHandle: this.handle, name: 'rev_reg_id' })
+  }
+
+  public get revocationRegistryIndex() {
+    const index = anoncreds.w3cCredentialGetAttribute({ objectHandle: this.handle, name: 'rev_reg_index' })
+    return index ? Number(index) : undefined
+  }
+
+  public toLegacy(): Credential {
+    return new Credential(
+      anoncreds.credentialFromW3C({
+        objectHandle: this.handle
+      }).handle
+    )
+  }
+
+  public static fromLegacy(options: W3CCredentialFromLegacyOptions): W3CCredential {
+    let credential
+    // Objects created within this method must be freed up
+    const objectHandles: ObjectHandle[] = []
+    try {
+      const credentialDefinition =
+        options.credentialDefinition instanceof CredentialDefinition
+          ? options.credentialDefinition.handle
+          : pushToArray(CredentialDefinition.fromJson(options.credentialDefinition).handle, objectHandles)
+
+      credential = new W3CCredential(
+        anoncreds.credentialToW3C({
+          objectHandle: options.credential.handle,
+          credentialDefinition
+        }).handle
+      )
+    } finally {
+      objectHandles.forEach((handle) => {
+        handle.clear()
+      })
+    }
+    return credential
+  }
+
+  public addNonAnonCredsIntegrityProof(proof: JsonObject) {
+    const credential = anoncreds.w3cCredentialAddNonAnonCredsIntegrityProof({
+      objectHandle: this.handle,
+      proof: JSON.stringify(proof)
+    })
+
+    this.handle.clear()
+    this.handle = credential
+  }
+
+  public setId(id: string) {
+    const credential = anoncreds.w3cCredentialSetId({
+      objectHandle: this.handle,
+      id
+    })
+
+    this.handle.clear()
+    this.handle = credential
+  }
+
+  public setSubjectId(id: string) {
+    const credential = anoncreds.w3cCredentialSetSubjectId({
+      objectHandle: this.handle,
+      id
+    })
+
+    this.handle.clear()
+    this.handle = credential
+  }
+
+  public addContext(context: string) {
+    const credential = anoncreds.w3cCredentialAddContext({
+      objectHandle: this.handle,
+      context
+    })
+
+    this.handle.clear()
+    this.handle = credential
+  }
+
+  public addType(type_: string) {
+    const credential = anoncreds.w3cCredentialAddType({
+      objectHandle: this.handle,
+      type_
```

Comment:
In JS it's fine to use `type` as a variable name I think. Is this taken over from the Python wrapper?

Would prefer to just use `type`, or otherwise if you want to avoid using `type`, use something like `credentialType`